### PR TITLE
Alerting: Add option to show inactive alerts in alert list panel

### LINKED ISSUE #76820: Alerting: Alert List Panel: setting the `Alert state filter` option to `Normal` doesn't return any results when using Prometheus data source
### What happened?

Alerting: Alert List Panel: setting the `Alert state filter` option to `Normal` doesn't return results when using Prometheus data source

Panel: https://play.grafana.org/d/Kb2RgVQnk/alert-panel?orgId=1&refresh=5s&editPanel=2

Screenshot 2023-10-19 at 3 10 59 pm

Screenshot 2023-10-19 at 3 02 00 pm

### What did you expect to happen?

The alert panel should display alerts in the `normal` (inactive) state i.e. non-firing or non-pending when the `Normal` option is enabled

### Did this work before?

Not sure

### How do we reproduce it?

1. Go to https://play.grafana.org/d/000000074/alerting?orgId=1&viewPanel=3&editPanel=3
2. Change the data source to `grafanacloud-play-prom` or any other Prometheus data source where there are alerts
3. In the Alert List Panel's setting change `Alert state filter` to `Normal` 

### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?

**Grafana:** Grafana v10.1.1 (0cfa76b22d)
**OS**: MacOS Ventura 13.6
**Browser:** Chrome 118.0.5993.70 

### Grafana platform?

Kubernetes

### Datasource(s)?

Mimir/Prometheus

### PR DESCRIPTION
**Pull request by bradleypettit**
**What is this feature?**

Adds an additional toggle to the Alert List panel, so that users can optionally display Prometheus alerts which have zero instances.

**Why do we need this feature?**

A number of users/customers have requested this feature. It also aligns more closely to the behavior of our Alert Rules page.

**Who is this feature for?**

Users who expect their data source managed alert rules to show in the Alert List panel when they are in the Normal state.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/76820

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Comment by bradleypettit** [comment]
Marked as a draft while we discuss if this approach is accepted. Once we're happy I'll add the new feature to the docs and look at adding a test case.

Example of new UI:
image

File: public/app/plugins/panel/alertlist/UnifiedAlertList.tsx
Comment (gillesdemey) [inline public/app/plugins/panel/alertlist/UnifiedAlertList.tsx:344]: Q: I'm a bit confused why we are also checking the `alertInstanceLabelFilter` here 🤔

Comment (bradleypettit) [reply public/app/plugins/panel/alertlist/UnifiedAlertList.tsx:344]: This is to basically give any instance label filters the user gave precedence over this setting. I tried to convey it in the FE for the user (below), but should have commented here too.
> If an Alert instance label filter is configured, inactive alert rules will be hidden

Most of the filters act on the alert rule itself; Alert name, datasource, folder. The one filter that acts on alert instances is `alertInstanceLabelFilter`, but seeing these inactive rules have zero instances, the filter had no effect. Here I check if the user provided any label filters, and if they did, hide the inactive alert rules.

**Comment by gillesdemey** [comment]
Approach looks good! Feel free to convert this from draft once you're ready for review / approval.

**Comment by bradleypettit** [comment]
Thanks @gillesdemey, appreciate the speedy feedback! Any tips on covering this use case in a test case? I'll try to do that and also add to the docs before I convert it.

File: docs/sources/panels-visualizations/visualizations/alert-list/index.md
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/alert-list/index.md:108]: Some edits to make this clearer and style guidelines
```suggestion
Filter for Prometheus alerts with no instances. Prometheus alerts have zero (0) instances when in the Normal (inactive) state and these are hidden by default. You can choose to show them by toggling this switch.
```

File: public/app/plugins/panel/alertlist/module.tsx
Comment (imatwawana) [inline public/app/plugins/panel/alertlist/module.tsx:155]: If there's more than one sentence in UI text, you have to include all periods. 

This UI text is a bit unclear. Do you mean that if you have the Alert instance label configured, it overrides this filter? If so, that should be said more explicitly and probably added in the docs.

```suggestion
        'Show inactive alert rules (for Prometheus alerts only). If the Alert instance label filter is configured, inactive alert rules will be hidden.',
```

**Review by imatwawana — COMMENTED** [review]
I've left some comments on UI text that could be clarified that might affect docs.

**Comment by bradleypettit** [comment]
> I've left some comments on UI text that could be clarified that might affect docs.

Thank you @imatwawana - I've updated with your suggestions. Let me know if you think some more changes are needed!

File: docs/sources/panels-visualizations/visualizations/alert-list/index.md
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/alert-list/index.md:108]: Just a couple final edits, but otherwise this looks good
```suggestion
Filter for Prometheus alerts with no instances. Prometheus alerts have zero (0) instances when in the Normal (inactive) state and these are hidden by default. You can choose to show them by toggling this switch. If the **Alert instance label** filter is configured, it takes precedence and inactive alert rules are hidden.
```

File: public/app/plugins/panel/alertlist/module.tsx
Comment (imatwawana) [inline public/app/plugins/panel/alertlist/module.tsx:155]:
```suggestion
        'Show inactive alert rules (for Prometheus alerts only). If the Alert instance label filter is configured, it takes precedence and inactive alert rules are hidden.',
```

**Review by imatwawana — APPROVED** [review]
Just taking a couple things out of future tense and putting them in present tense, but otherwise LGTM!

File: docs/sources/panels-visualizations/visualizations/alert-list/index.md
Comment (brendamuir) [inline docs/sources/panels-visualizations/visualizations/alert-list/index.md:108]: Toggle to show alert rules with no alert instances. 

Alert rules have zero (0) alert instances when in **Normal**(inactive) state and are hidden by default.

Comment (brendamuir) [reply docs/sources/panels-visualizations/visualizations/alert-list/index.md:108]: I am not sure what this means? If you have configured the **Alert instance label** filter, it takes precedence and inactive alert rules are hidden.

what does it take precedence over? If you've configured the filter then the toggle doesn't work?

File: public/app/plugins/panel/alertlist/module.tsx
Comment (brendamuir) [reply public/app/plugins/panel/alertlist/module.tsx]: what do you mean by prometheus alert rules as opposed to other alert rule types?

File: docs/sources/panels-visualizations/visualizations/alert-list/index.md
Comment (bradleypettit) [reply docs/sources/panels-visualizations/visualizations/alert-list/index.md:108]: This comment above might explain the logic behind it a little more too - https://github.com/grafana/grafana/pull/96888#discussion_r1854630925

Ultimately I trialled both behaviors here (one where we ignore the instance label filter and this one, where we honor it). Considering the other filters again now, there is already some level of precedence which is implied; for example if someone filters by an alert rule name, and then also applies an instance label filter, the instance label filter will win.

In this case it might be over complicating things to explain that how it reacts to the Alert instance label filter?

File: public/app/plugins/panel/alertlist/module.tsx
Comment (bradleypettit) [reply public/app/plugins/panel/alertlist/module.tsx:155]: Made a couple of new commits that invalidate this question, see latest comments.

**Comment by bradleypettit** [comment]
I tested this again and overlooked that Grafana Managed alerts can have the same behavior if they are written like Prometheus alerts (with the condition as part of the query). So I've removed any mention of Grafana/Prometheus based rules as it applied to both.

I tried to reword the text mentioning the instance label filter to be a little less confusing.

I think there could be an argument for showing these to be the default behavior, however doing so would be a breaking change for people using the panel already. So I feel introducing this switch is still the right path because of that.

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM! :shipit:

**Comment by bradleypettit** [comment]
Mistakenly rebased 😭  - ignore approval requests for now.

**Comment by bradleypettit** [comment]
Apologies for the spam - was fixing up some doc conflicts in main and I mistakenly pushed a rebase... all sorted now and ready to merge (I don't have access to do so myself).

@imatwawana - just tagging you in case you want to double check the docs page given your recent refactor.

File: docs/sources/panels-visualizations/visualizations/alert-list/index.md
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/alert-list/index.md:108]:
```suggestion
| Show alerts with 0 instances | Filter for alert rules with no instances. Alert rules with 0 (zero) instances are hidden by default. You can choose to show them by toggling this switch. Because these rules have no instances, they remain hidden if the **Alert instance label** filter is configured. |
```

File: public/app/plugins/panel/alertlist/module.tsx
Comment (imatwawana) [inline public/app/plugins/panel/alertlist/module.tsx:155]:
```suggestion
        'Include alert rules which have 0 (zero) instances. Because these rules have no instances, they remain hidden if the Alert instance label filter is configured.',
```

**Review by imatwawana — APPROVED** [review]
Thanks for the heads up. I don't see any issues with regard to the refactor. I made a couple small wording edits. If those work for you, then I'll go ahead and merge this once those changes are committed.

**Comment by bradleypettit** [comment]
> Thanks for the heads up. I don't see any issues with regard to the refactor. I made a couple small wording edits. If those work for you, then I'll go ahead and merge this once those changes are committed.

Thanks @imatwawana for the final tweaks - good to go now.

**Pull request merged**
